pkg/framework/capi-machinesets.go

```diff
@@ -0,0 +1,195 @@
+	Expect(msName).ToNot(BeEmpty())
+	Expect(clusterName).ToNot(BeEmpty())
+	Expect(infrastructureRef.APIVersion).ToNot(BeEmpty())
+	Expect(infrastructureRef.Kind).ToNot(BeEmpty())
+	Expect(infrastructureRef.Name).ToNot(BeEmpty())
+
+	return CAPIMachineSetParams{
+		msName:            msName,
+		clusterName:       clusterName,
+		replicas:          replicas,
+		infrastructureRef: infrastructureRef,
+		failureDomain:     failureDomain,
+	}
+}
+
+// CreateCAPIMachineSet creates a new MachineSet resource.
+func CreateCAPIMachineSet(ctx context.Context, cl client.Client, params CAPIMachineSetParams) (*clusterv1.MachineSet, error) {
+	By(fmt.Sprintf("Creating MachineSet %q", params.msName))
+
+	userDataSecret := "worker-user-data"
+
+	ms := &clusterv1.MachineSet{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "MachineSet",
+			APIVersion: "machine.openshift.io/v1beta1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      params.msName,
+			Namespace: ClusterAPINamespace,
+		},
+		Spec: clusterv1.MachineSetSpec{
+			Replicas:    &params.replicas,
+			ClusterName: params.clusterName,
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"machine.openshift.io/cluster-api-cluster":    params.clusterName,
+					"machine.openshift.io/cluster-api-machineset": params.msName,
+				},
+			},
+			Template: clusterv1.MachineTemplateSpec{
+				ObjectMeta: clusterv1.ObjectMeta{
+					Labels: map[string]string{
+						"machine.openshift.io/cluster-api-cluster":    params.clusterName,
+						"machine.openshift.io/cluster-api-machineset": params.msName,
+					},
+				},
+				Spec: clusterv1.MachineSpec{
+					Bootstrap: clusterv1.Bootstrap{
+						DataSecretName: &userDataSecret,
+					},
+					ClusterName:       params.clusterName,
+					InfrastructureRef: params.infrastructureRef,
+				},
+			},
+		},
+	}
+
+	if params.failureDomain != "" {
+		ms.Spec.Template.Spec.FailureDomain = &params.failureDomain
+	}
+
+	Expect(cl.Create(ctx, ms)).To(Succeed())
+
+	return ms, nil
+}
+
+// WaitForCAPIMachineSetsDeleted polls until the given MachineSets are not found, and
+// there are zero Machines found matching the MachineSet's label selector.
+func WaitForCAPIMachineSetsDeleted(ctx context.Context, cl client.Client, machineSets ...*clusterv1.MachineSet) {
+	for _, ms := range machineSets {
+		By(fmt.Sprintf("Waiting for MachineSet %q to be deleted", ms.GetName()))
+		Eventually(func() bool {
+			selector := ms.Spec.Selector
+
+			machines, err := GetCAPIMachines(ctx, cl, &selector)
+			if err != nil || len(machines) != 0 {
+				return false // Still have Machines, or other error.
+			}
+
+			err = cl.Get(ctx, client.ObjectKey{
+				Name:      ms.GetName(),
+				Namespace: ms.GetNamespace(),
+			}, &clusterv1.MachineSet{})
+
+			return apierrors.IsNotFound(err) // MachineSet and Machines were deleted.
+		}, WaitLong, RetryMedium).Should(BeTrue())
```

Comment:
With all the `.Should()` and `.To()` (both with the Eventuallys and the Expects) could we add the optional description string like `Eventually(...).Should(BeTrue(), "it should have been able to delete all the CAPI MachineSets")`?
That helps in debugging when ginkgo reports a failure. Thanks!